File: src/main/java/mediathek/tool/notification/LinuxNotificationCenter.java
package mediathek.tool.notification;

import es.blackleg.jlibnotify.JLibnotify;
import es.blackleg.jlibnotify.core.DefaultJLibnotifyLoader;
import es.blackleg.jlibnotify.exception.JLibnotifyInitException;
import es.blackleg.jlibnotify.exception.JLibnotifyLoadException;
import mediathek.config.Config;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.Closeable;

public class LinuxNotificationCenter implements INotificationCenter, Closeable {
    private static final Logger logger = LogManager.getLogger();
    private JLibnotify libNotify;
    private boolean nativeSupport;

    public LinuxNotificationCenter() {
        try {
            var loader = new DefaultJLibnotifyLoader();
            libNotify = loader.load();
            libNotify.init("MediathekView");
            nativeSupport = true;

            var serverInfo = libNotify.getServerInfo();
            logger.info(serverInfo);

            if (Config.isDebugModeEnabled()) {
                logger.debug("Server capabilities:");
                var caps = libNotify.getServerCapabilities();
                for (var s : caps)
                    logger.debug("\t {}", s);
            }
        }
        catch (UnsatisfiedLinkError | RuntimeException | JLibnotifyLoadException | JLibnotifyInitException e) {
            nativeSupport = false;
            logger.error("Failed to initialize libNotify",e);
        }
    }

    @Override
    public void displayNotification(NotificationMessage msg) {
        var notification = libNotify.createNotification(msg.getTitle(), msg.message,"dialog-information");
        notification.show();
    }

    public boolean hasNativeSupport() {
        return nativeSupport;
    }

    @Override
    public void close() {
        libNotify.unInit();
    }
}

File: src/main/kotlin/mediathek/tool/notification/LinuxNotificationCenter.kt
package mediathek.tool.notification

import es.blackleg.jlibnotify.JLibnotify
import es.blackleg.jlibnotify.core.DefaultJLibnotifyLoader
import mediathek.config.CommandLineOptions
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

class LinuxNotificationCenter private constructor(
    private val libNotify: JLibnotify,
) : NotificationBackend {
    private val lifecycleLock = Any()
    private var closed = false

    init {
        logger.info(libNotify.serverInfo)

        if (CommandLineOptions.isDebugModeEnabled()) {
            logger.debug("Server capabilities:")
            for (capability in libNotify.serverCapabilities) {
                logger.debug("\t {}", capability)
            }
        }
    }

    override fun publish(notification: NotificationMessage) {
        synchronized(lifecycleLock) {
            if (closed) {
                return
            }
            val nativeNotification =
                libNotify.createNotification(notification.title, notification.message, "dialog-information")
            nativeNotification.show()
        }
    }

    override fun close() {
        synchronized(lifecycleLock) {
            if (closed) {
                return
            }
            closed = true
            libNotify.unInit()
        }
    }

    companion object {
        private val logger: Logger = LogManager.getLogger()

        fun create(): LinuxNotificationCenter {
            val libNotify = DefaultJLibnotifyLoader().load()
            libNotify.init("MediathekView")
            return try {
                LinuxNotificationCenter(libNotify)
            } catch (error: Throwable) {
                try {
                    libNotify.unInit()
                } catch (closeError: RuntimeException) {
                    error.addSuppressed(closeError)
                }
                throw error
            }
        }
    }
}
